### PR TITLE
[WIP] Add multi-gpu support

### PR DESCRIPTION
**Pull request by nottombrown**
This should be merged after #9 

Adds multi-gpu support by using the data-parallel pytorch wrapper. 

This currently completes 3 cycles in ~2 minutes using 4x V100 GPUs, and results in somewhat worse final performance. I haven't done any hyperparam tuning.

## Running on 1 GPU
```
time CUDA_VISIBLE_DEVICES=0 python train.py --dataset rocstories  --analysis --data_dir cloze_data --n_batch=20 --n_gpu=1 --log_dir=logs --desc simple-datapar
```

```
cat logs/simple-datapar.jsonl | jq
{
  "desc": "simple-datapar",
  "dataset": "rocstories",
  "log_dir": "logs",
  "save_dir": "save/",
  "data_dir": "cloze_data",
  "submission_dir": "submission/",
  "submit": false,
  "analysis": true,
  "seed": 42,
  "n_iter": 3,
  "n_batch": 20,
  "max_grad_norm": 1,
  "lr": 6.25e-05,
  "lr_warmup": 0.002,
  "n_ctx": 512,
  "n_embd": 768,
  "n_head": 12,
  "n_layer": 12,
  "embd_pdrop": 0.1,
  "attn_pdrop": 0.1,
  "resid_pdrop": 0.1,
  "clf_pdrop": 0.1,
  "l2": 0.01,
  "vector_l2": false,
  "n_gpu": 1,
  "opt": "adam",
  "afn": "gelu",
  "lr_schedule": "warmup_linear",
  "encoder_path": "model/encoder_bpe_40000.json",
  "bpe_path": "model/vocab_40000.bpe",
  "n_transfer": 12,
  "lm_coef": 0.5,
  "b1": 0.9,
  "b2": 0.999,
  "e": 1e-08,
  "n_valid": 374,
  "time": 1530070189.6090798
}
{
  "n_epochs": 1,
  "n_updates": 74,
  "tr_cost": 36.685862974687055,
  "va_cost": 48.44379649187792,
  "tr_acc": 86.09625668449198,
  "va_acc": 82.0855614973262,
  "time": 1530070278.9061887
}
{
  "n_epochs": 2,
  "n_updates": 148,
  "tr_cost": 34.50088770249311,
  "va_cost": 71.0536474977585,
  "tr_acc": 92.7807486631016,
  "va_acc": 85.8288770053476,
  "time": 1530070347.4885302
}
{
  "n_epochs": 3,
  "n_updates": 222,
  "tr_cost": 10.390687029629468,
  "va_cost": 111.5951719233059,
  "tr_acc": 96.52406417112299,
  "va_acc": 87.70053475935828,
  "time": 1530070414.4696817
}
```

## Running on four GPUS results in somewhat worse performance
```
time CUDA_VISIBLE_DEVICES=0,1,2,3 python train.py --dataset rocstories  --analysis --data_dir cloze_data --n_batch=16 --n_gpu=4 --log_dir=logs --desc simple-datapar-4gpu
```
```
{
  "desc": "simple-datapar-4gpu",
  "dataset": "rocstories",
  "log_dir": "logs",
  "save_dir": "save/",
  "data_dir": "cloze_data",
  "submission_dir": "submission/",
  "submit": false,
  "analysis": true,
  "seed": 42,
  "n_iter": 3,
  "n_batch": 16,
  "max_grad_norm": 1,
  "lr": 6.25e-05,
  "lr_warmup": 0.002,
  "n_ctx": 512,
  "n_embd": 768,
  "n_head": 12,
  "n_layer": 12,
  "embd_pdrop": 0.1,
  "attn_pdrop": 0.1,
  "resid_pdrop": 0.1,
  "clf_pdrop": 0.1,
  "l2": 0.01,
  "vector_l2": false,
  "n_gpu": 4,
  "opt": "adam",
  "afn": "gelu",
  "lr_schedule": "warmup_linear",
  "encoder_path": "model/encoder_bpe_40000.json",
  "bpe_path": "model/vocab_40000.bpe",
  "n_transfer": 12,
  "lm_coef": 0.5,
  "b1": 0.9,
  "b2": 0.999,
  "e": 1e-08,
  "n_valid": 374,
  "time": 1530070504.329161
}
{
  "n_epochs": 1,
  "n_updates": 23,
  "tr_cost": 456.34622012867646,
  "va_cost": 492.78721643131684,
  "tr_acc": 77.27272727272727,
  "va_acc": 75.93582887700535,
  "time": 1530070587.3708036
}
{
  "n_epochs": 2,
  "n_updates": 46,
  "tr_cost": 331.6804982557654,
  "va_cost": 717.7863782586899,
  "tr_acc": 87.96791443850267,
  "va_acc": 82.88770053475936,
  "time": 1530070615.6105187
}
{
  "n_epochs": 3,
  "n_updates": 69,
  "tr_cost": 436.451277625752,
  "va_cost": 723.3510768298797,
  "tr_acc": 90.10695187165776,
  "va_acc": 84.75935828877004,
  "time": 1530070645.8095183
}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -60,6 +61,3 @@ The single run test accuracy of this PyTorch version is 85.84%, while the author
 The authors implementations uses 8 GPU and can thus accomodate a batch of 64 samples while the present implementation is single GPU and is in consequence limited to 20 instances on a K80 for memory reasons. In our test, increasing the batch size from 8 to 20 samples increased the test accuracy by 2.5 points. A better accuracy may be obtained by using a multi-GPU setting (on the TO-DO list).
```
Comment (thomwolf) [inline README.md]: Can you also remove this TODO mention related to multi-GPU and maybe add a mention of your single-run test?

File: train.py [hunk 2]
```diff
@@ -212,37 +213,48 @@ def run_epoch():
 
     args = parser.parse_args()
     print(args)
-    globals().update(args.__dict__) #TODO maybe we want to remove these gobal variables to make it cleaner
-    random.seed(seed)
-    np.random.seed(seed)
-    torch.manual_seed(seed)
-    torch.cuda.manual_seed_all(seed)
+    # globals().update(args.__dict__)  # TODO maybe we want to remove these gobal variables to make it cleaner
```
Comment (thomwolf) [inline train.py]: We can remove this also

File: train.py [hunk 3]
```diff
@@ -197,7 +198,7 @@ def run_epoch():
     parser.add_argument('--clf_pdrop', type=float, default=0.1)
     parser.add_argument('--l2', type=float, default=0.01)
     parser.add_argument('--vector_l2', action='store_true')
-    parser.add_argument('--n_gpu', type=int, default=1)#4) # TODO add mutli-gpu training logic
+    parser.add_argument('--n_gpu', type=int, default=1)  # 4) # TODO add mutli-gpu training logic
```
Comment (thomwolf) [inline train.py]: And this TODO comment

**Comment by nottombrown** [comment]
Feel free to make those changes directly and merge it in if you'd like! I'm
likely not going to have a chance to work on this more for a couple weeks.

**Comment by thomwolf** [comment]
Ok no problem, I'll include them in a few improvements I'll push next week. Thanks for that!

**Pull request merged**